packages/credential-provider-assume-role/src/fromTokenFile.ts

```diff
@@ -0,0 +1,126 @@
+
+export interface FromTokenFileInit extends SourceProfileInit {
+  /**
+   * A function that assumes a role with web identity and returns a promise fulfilled with
+   * credentials for the assumed role.
+   *
+   * @param sourceCreds The credentials with which to assume a role.
+   * @param params
+   */
+  roleAssumerWithWebIdentity: (params: AssumeRoleWithWebIdentityCommandInput) => Promise<Credentials>;
+  /**
+   * A function that assumes a role and returns a promise fulfilled with
+   * credentials for the assumed role.
+   *
+   * @param sourceCreds The credentials with which to assume a role.
+   * @param params
+   */
+  roleAssumer?: (sourceCreds: Credentials, params: AssumeRoleCommandInput) => Promise<Credentials>;
+}
+
+/**
+ * Represents OIDC credentials from a file on disk.
+ */
+export const fromTokenFile = (init: FromTokenFileInit): CredentialProvider => async () => {
+  if (process.env[ENV_TOKEN_FILE] && process.env[ENV_ROLE_ARN]) {
+    return resolveCredentialsFromEnv(init);
+  }
+  const profiles = await parseKnownFiles(init);
+  return resolveCredentialsFromIni(getMasterProfileName(init), profiles, init);
+};
+
+const resolveCredentialsFromEnv = async (options: FromTokenFileInit) => {
+  if (!options.roleAssumerWithWebIdentity) {
+    throw new ProviderError(
+      `Role Arn '${process.env[ENV_ROLE_ARN]}' needs to be assumed with web identity,` +
+        ` but no role assumption callback was provided.`,
+      false
+    );
+  }
+  return options.roleAssumerWithWebIdentity({
+    WebIdentityToken: readFileSync(process.env[ENV_TOKEN_FILE]!, { encoding: "ascii" }),
+    RoleArn: process.env[ENV_ROLE_ARN],
+    RoleSessionName: process.env[ENV_ROLE_SESSION_NAME] || `aws-sdk-js-session-${Date.now()}`,
+  });
+};
+
+const resolveCredentialsFromIni = async (
+  profileName: string,
+  profiles: ParsedIniData,
+  options: FromTokenFileInit,
+  visitedProfiles: { [profileName: string]: true } = {}
+): Promise<Credentials> => {
+  const data = profiles[profileName];
+
+  if (isAssumeRoleSourceProfile(data)) {
+    const {
+      role_arn: RoleArn,
+      role_session_name: RoleSessionName = `aws-sdk-js-session-${Date.now()}`,
+      source_profile,
+    } = data;
+
+    if (!options.roleAssumer) {
+      throw new ProviderError(
+        `Profile '${profileName}' requires a role to be assumed,` + ` but no role assumption callback was provided.`,
+        false
+      );
+    }
+
+    if (source_profile in visitedProfiles) {
+      throw new ProviderError(
+        `Detected a cycle attempting to resolve credentials for profile` +
+          ` ${getMasterProfileName(options)}. Profiles visited: ` +
+          Object.keys(visitedProfiles).join(", "),
+        false
+      );
+    }
+
+    const sourceCreds = resolveCredentialsFromIni(source_profile, profiles, options, {
+      ...visitedProfiles,
+      [source_profile]: true,
+    });
+
+    return options.roleAssumer(await sourceCreds, { RoleArn, RoleSessionName });
```

Comment:
If we allow roleAssumer chaining here, should we allow other assume role options? Like MFA serial, and MFA token? I'm comparing to https://github.com/aws/aws-sdk-js-v3/blob/34cecf1e3b4048dfdf88422e369615843d798c81/packages/credential-provider-ini/src/index.ts#L188-L197

Reply:
> If we allow roleAssumer chaining here, should we allow other assume role options?

Short answer: Yes.
Long answer: Not in this PR. A separate PR should be created to introduce reusable `fromAssumeRole` function.